webpack.plugin.localize.js

```diff
@@ -0,0 +1,166 @@
+
+const ConstDependency = require("webpack/lib/dependencies/ConstDependency");
+const fs = require("node:fs");
+const {resolve} = require("node:path");
+const path = require("path");
+const gettextParser = require("gettext-parser");
+
+// generate and then load the locale translation data
+const baseDir = __dirname;
+const localeDir = path.resolve(baseDir, "warehouse/locale");
+const KNOWN_LOCALES = [
+  "en",  // English
+  "es",  // Spanish
+  "fr",  // French
+  "ja",  // Japanese
+  "pt_BR",  // Brazilian Portuguese
+  "uk",  // Ukrainian
+  "el",  // Greek
+  "de",  // German
+  "zh_Hans",  // Simplified Chinese
+  "zh_Hant",  // Traditional Chinese
+  "ru",  // Russian
+  "he",  // Hebrew
+  "eo",  // Esperanto
+];
+
+// A custom regular expression to do some basic checking of the plural form,
+// to try to ensure the plural form expression contains only expected characters.
+// - the plural form expression MUST NOT have any type of quotes and
+//   the only whitespace allowed is space (not tab or form feed)
+// - MUST NOT allow brackets other than parentheses (()),
+//   as allowing braces ({}) might allow ending the function early
+// - MUST allow space, number variable (n), numbers, groups (()),
+//   comparisons (<>!=), ternary expressions (?:), and/or (&|),
+//   remainder (%)
+const pluralFormPattern = new RegExp("^ *nplurals *= *[0-9]+ *; *plural *=[ n0-9()<>!=?:&|%]+;?$");
+
+const allLocaleData = KNOWN_LOCALES
+  .filter(langCode => langCode !== "en")
+  .map((langCode) => resolve(localeDir, langCode, "LC_MESSAGES/messages.po"))
+  .filter((file) => fs.statSync(file).isFile())
+  .map((file) => ({path: path.relative(baseDir, file), data: fs.readFileSync(file, "utf8")}))
+  .map((data) => {
+    try {
+      const lines = data.data
+        .split("\n")
+        // gettext-parser does not support obsolete previous translations,
+        // so filter out those lines
+        // see: https://github.com/smhg/gettext-parser/issues/79
+        .filter(line => !line.startsWith("#~|"))
+        .join("\n");
+      const parsed = gettextParser.po.parse(lines);
+      const language = parsed.headers["Language"];
+      const pluralForms = parsed.headers["Plural-Forms"];
+      const result = {
+        "": {
+          "language": language,
+          "plural-forms": pluralForms,
+        },
+      };
+
+      if (!pluralFormPattern.test(pluralForms)) {
+        throw new Error(`Invalid plural forms for '${language}': "${pluralForms}"`);
+      }
+
+      const translations = parsed.translations[""];
+      for (const key in translations) {
+        if (key === "") {
+          continue;
+        }
+        const value = translations[key];
+        const refs = value.comments.reference.split("\n");
+        if (refs.every(refLine => !refLine.includes(".js:"))) {
+          continue;
+        }
+        // The 'msgid' and 'msgstr' must be strings.
+        result[value.msgid.toString()] = value.msgstr.toString();
```

Comment:
I think we need to take this a step further, because these values will also be embedded directly into the UI, sometimes as attributes on DOM elements. I think we would be vulnerable to XSS and script injection that could embedded in a translation string. Most places we use `textContent` which protects us, but we don't have any guarantees that all translated strings in JS are injected via `textContent`. I think this might need to be something like:

```suggestion
        result[value.msgid] = value.msgstr
            .replace(/&/g, "&amp;")
            .replace(/</g, "&lt;")
            .replace(/>/g, "&gt;")
            .replace(/"/g, "&quot;")
            .replace(/'/g, "&#39;");
```
I think I'm missing what `toString` would be protecting us from here, can you explain?

Reply:
Ah, I was assuming that `.textContent` was always used. In that case, your suggestion makes sense.